core/lib/src/fairing/ad_hoc.rs

```diff
@@ -66,10 +66,10 @@ impl AdHoc {
     /// use rocket::fairing::AdHoc;
     ///
     /// // The no-op attach fairing.
-    /// let fairing = AdHoc::on_attach("No-Op", |rocket| Ok(rocket));
+    /// let fairing = AdHoc::on_attach("No-Op", |rocket| Box::pin(async move { Ok(rocket) }));
     /// ```
     pub fn on_attach<F: Send + 'static>(name: &'static str, f: F) -> AdHoc
-        where F: FnOnce(Rocket) -> Result<Rocket, Rocket>
+        where F: FnOnce(Rocket) -> BoxFuture<'static, Result<Rocket, Rocket>>
```

Comment:
Yeah, it should.

It might be nice to do that for `on_request` and `on_response` too, but the lack of support for `for<'a> Fn(&'a Request) -> Fut, Fut: 'a` may make it harder or impossible for those particular functions.